File: Chapter-3/BST_DS.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


class Binary_search_tree:
    def Create_Tree(self, root, temp):
        new_node = Node(temp)
        new_node.left = None
        new_node.right = None
        return new_node

    def Insert_Tree(self, root, temp):
        if root == None:
            return self.Create_Tree(root, temp)
        elif root.value > temp:
            root.left = self.Insert_Tree(root.left, temp)
        else:
            root.right = self.Insert_Tree(root.right, temp)
        return root

    def successor(self, root):
        root = root.right
        while root.left != None:
            root = root.left
        return root

    def predessor(self, root):
        root = root.left
        while (root.right != None):
            root = root.right
        return root

    def Delete_Tree(self, root, data):
        if root is None:
            return root
        if root.value > data:
            root.left = self.Delete_Tree(root.left, data)
        elif root.value < data:
            root.right = self.Delete_Tree(root.right, data)
        else:
            if root.left is None and root.right is None:
                return None
            elif root.left is not None:
                pre = self.predessor(root)
                root.value = pre.value
                root.left = self.Delete_Tree(root.left, root.value)
            else:
                pre = self.successor(root)
                root.value = pre.value
                root.right = self.Delete_Tree(root.right, root.value)
        return root

    def searching(self, root, temp):
        if root is None:
            return None
        if root.value == temp:
            return root
        elif root.value > temp:
            return self.searching(root.left, temp)
        else:
            return self.searching(root.right, temp)

    def Print(self, root):
        if root is None:
            return
        self.Print(root.left)
        print(root.value, end=" ")
        self.Print(root.right)


root = None
tree = Binary_search_tree()
root = tree.Insert_Tree(root, 39)
root = tree.Insert_Tree(root, 29)
root = tree.Insert_Tree(root, 1)
root = tree.Insert_Tree(root, 87)
root = tree.Insert_Tree(root, 63)
tree.Print(root)
print()

print("after deletion ")
root = tree.Delete_Tree(root, 63)
tree.Print(root)
print()

print("after searching 39")
n = tree.searching(root, 1)
if (n != None):
    print(f"{n.value} Found")
else:
    print("not present in the tree")

File: Chapter-3/BST_DS.cpp
#include<bits/stdc++.h>

using namespace std;

struct Node {
    int value;
    struct Node *left;
    struct Node *right;
};

struct Node * Create_Tree(int temp){
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    newNode->value = temp;
    newNode->left = NULL;
    newNode->right = NULL;
    return newNode;
};

struct Node * Insert_Tree(struct Node* root,int temp){
    if(root==NULL){
        return Create_Tree(temp);
    }
    else if(root->value > temp){
        root->left = Insert_Tree(root->left,temp);
    }
    else{
        root->right = Insert_Tree(root->right,temp);
    }
    return root;
};

struct Node *inOrderPredecessor(struct Node* root){
    root = root->left;
    while (root->right!=NULL){
        root = root->right;
    }
    return root;
}
struct Node *inOrderSuccessor(struct Node* root){
    root = root->right;
    while (root->left!=NULL){
        root = root->left;
    }
    return root;
}

struct Node *delete_Node(struct Node *root, int value){
    struct Node* iPre;
    if (root == NULL){
        cout<<"Element is not Present in Tree"<<endl;
        return root;
    }
    if (value < root->value){
        root-> left = delete_Node(root->left,value);
    }
    else if (value > root->value){
        root-> right = delete_Node(root->right,value);
    }
    else{
        if (root->left==NULL&&root->right==NULL){
        free(root);
        return NULL;
        }
        else if(root->right != NULL){
            iPre = inOrderSuccessor(root);
            root->value = iPre->value;
            root->right = delete_Node(root->right,root->value);
        }
        else{
            iPre = inOrderPredecessor(root);
            root->value = iPre->value;
            root->left = delete_Node(root->left,root->value);
        }
    }
    return root;
}

struct Node *searching(struct Node *root,int temp){
    if(root==NULL){
        return NULL;
    }
    if(root->value == temp){
        return root;
    }
    else if(root->value>temp){
        return searching(root->left,temp);
    }
    else{
        return searching(root->right,temp);
    }
};


void in_order(struct Node *root){
    if(root==NULL){
        return;
    }
    in_order(root->left);
    cout<<root->value<<" ";
    in_order(root->right);
}

int main(){
    struct Node *root = NULL;
    struct Node *n;

  int task;
    do{
        cout<<"1. insertion in Tree"<<endl;
        cout<<"2. searching in a Tree"<<endl;
        cout<<"3. deletion in Tree"<<endl;
        cout<<"4. display The Tree"<<endl;
        cout<<"0. for Exit"<<endl;
        cout<<"enter a tsk to be performed : ";
        cin>>task;
        switch(task){
        case 0:
            break;
        case 1:
            int value;
            cout<<"enter the value: ";
            cin>>value;
            root = Insert_Tree(root,value);
            cout<<endl;
            break;

        case 2:
            int temp;
            cout<<"enter the value for Searching : ";
            cin>>temp;
            n = searching(root,temp);
            if(n!=NULL){
                cout<<"Found: "<<n->value<<endl;
            }
            else{
                cout<<temp<<" not present in the tree"<<endl;
            }
            cout<<endl;
            break;

        case 3:

            cout<<"enter the value for deletion : ";
            cin>>value;
            root = delete_Node(root,value);
            cout<<endl;
            break;

        case 4:
            in_order(root);
            cout<<endl;
            break;

        default:
            cout<<"please enter correct choice"<<endl;
            cout<<endl;
            break;
        }
    }while(task!=0);

    return 0;
}
